src/peft/tuners/trainable_tokens/layer.py

```diff
@@ -37,24 +37,40 @@ def __init__(
         base_layer: nn.Module,
         adapter_name: str,
         token_indices: list[int],
+        tied_adapter: Optional[TrainableTokensLayer] = None,
         **kwargs,
     ) -> None:
         super().__init__()
 
         self.base_layer = base_layer
         self._active_adapter = adapter_name
-        self.token_indices = {}
         self.kwargs = kwargs
 
+        self.tied_adapter = tied_adapter
+
         # we store the updated weights of particular tokens and their originals. we assume
         # that the count of new tokens is far smaller than the number of total tokens.
-        self.trainable_tokens_delta = nn.ParameterDict({})
-        self.trainable_tokens_original = BufferDict({})
+        #
+        # In case we have weight tying with another token adapter, we'll have no actual
+        # references on our own but use everything from the tied adapter.
+        if not self.tied_adapter:
+            self.trainable_tokens_delta = nn.ParameterDict({})
+            self.trainable_tokens_original = BufferDict({})
+            self.token_indices = {}
+        else:
+            self.trainable_tokens_delta = self.tied_adapter.trainable_tokens_delta
```

Comment:
IIRC, you mentioned that stuff like `model.state_dict()` already correctly works for this case thanks to transformers. Could you add a comment for that? When the model is not a transformers model, would we get the same param twice?

Reply:
It turns out it is more subtle than that. While transformers keeps an eye out for tied weights on saving/loading we still have to deal with the tied weights we produce by using the tied adapter. I've added a test and a solution (wrapping the tied adapter in a list so that it is excluded from `.*modules()` calls).

My first idea was to wrap the parameters from the tied adapter in a `BufferDict` which turned out to be unnecessary with the list wrapping solution. That's the reason why there are now some fixes + tests for the buffer dict in this PR as well...